File: 15_lattice-paths.py
#!/usr/bin/python

count = 0
cache = {}

def path(i, j, N):
	global count
	global cache

	if (i == N and j == N):
		count += 1
		return
	
	if (i, j) in cache:
		print "in cache: ", (i, j)
		count += 1
		return
	else:
		cache[(i, j)] = 42
		print (i, j)
	
	if (i == N):
		path(i, j+1, N)
	elif (j == N):
		path(i+1, j, N)
	else:
		path(i+1, j, N)
		path(i, j+1, N)

N = int(raw_input("Enter size of grid: "))
path(0, 0, N)
print count

File: 15_lattice-paths.c
#include <stdio.h>

static int count = 0;

void path(int mat[21][21], int N, int i, int j)
{
	if (i == N && j == N) {
		count++;
		return;
	}
	/*for (int i = 0; i < 3; i++) {
		for (int j = 0; j < 3; j++) {
			printf("[%d] ", mat[i][j]);
		}
		printf("\n");
	}*/
	
	if (i == N) {
		path(mat, N, i, j+1);
	} else if (j == N) {
		path(mat, N, i+1, j);
	} else {
		path(mat, N, i+1, j);
		path(mat, N, i, j+1);
	}
	

}

int main()
{
	int mat[21][21] = { 0 };

	path(mat, 20, 0, 0);

	printf("count: %d\n", count);

	return 0;
}
